Introduce immutable string, array and map

cecton opened the pull request:
#### Description

Please check the example `immutable` in this PR to see how it is used.

I have started some time ago an experiment of porting [Immutable.js](https://immutable-js.com/) to Yew. This basically leverage the use of `ImplicitClone` on 3 core types: `String`, `Array` and `Map`. And use garbage collection with Rc.

- The immutable string type already exists in Yew for some time. It has been introduced in #1994 It's very handy for passing cheap strings around and it is also very useful for library maintainers who want to have string properties that work with `&str` and `String` indifferently and with Rc.
- The immutable array type allows creating a cheap Rc'ed array of static size of any type. It's very similar to the immutable string type.
- The immutable map type allows creating a cheap IndexMap. This is slightly more complicated because the map is not a primitive in Rust. Here we use either: an `Rc<IndexMap<K, V>>` or a `&'static [(K, V)]`.

Following the discussion in #2448 I realized that it is better to create a more generic "immutable" crate rather than crating a "yew-immutable" crate. This is why I created this new project: https://github.com/rustminded/imut It's currently not published on crates.io yet because I prefer getting the green light here first. If my PR here is approved I will add the documentation and publish `imut`. (I couldn't take the crate `immutable` because it is already taken and I couldn't reach the owner.)

To avoid breaking compatibility I added a type alias from `AttrValue` to `IString`. Those types were identical anyway.

I also added a re-export of `imut` to `yew::immutable` so people can import them easily if they want without the need to add `imut` to their dependencies.

It is best for the reviewers to also review https://github.com/rustminded/imut and put their remarks in this PR. The most important changes are there after all.

Fixes #2448 

#### Checklist

<!-- For further details, please read CONTRIBUTING.md -->

- [x] I have run `cargo make pr-flow`
- [x] I have reviewed my own code
- [x] I have added tests
  <!-- If this is a bug fix, these tests will fail if the bug is present (to stop it from cropping up again) -->
  <!-- If this is a feature, my tests prove that the feature works -->
- [ ] I have added documentation to `imut`
- [ ] I have added a similar API of immutable.js to `imut`
- [ ] I have published `imut`

cecton commented:
:wave: @hamza1311 can I have your feedback on this before I complete the documentation and publish `imut`? Thanks

hamza1311 commented:
This looks fine for the most part. `ImplicitClone` should still be part of Yew. `immutable` doesn't use it whereas Yew does. It may also help us avoid issues with orphan rules down the line where we can't implement `ImplicitClone` for a type because we don't own the trait

cecton commented:
> `ImplicitClone` should still be part of Yew.

Actually that's not good. `ImplicitClone` means it's cheap to clone. From Yew's perspective it means we can pass it easily through props. But from imut's perspective it means the type doesn't need to be taken by reference but can be cloned easily.

This is why in IArray, for instance, there is a requirement that T implements ImplicitClone so when you call `.iter()` on IArray it yields owned types and not references. If you check the yew example I made here for IArray, in the method `fn update()` of the yew component, you will see that it is very convenient.

> It may also help us avoid issues with orphan rules down the line where we can't implement `ImplicitClone` for a type because we don't own the trait

That's a Rust problem in general. But I think this is a special case. Usually it is solved by wrapping the type in a container and implementing Deref. But I can also imagine adding features to `imut` that would implement that trait for specific crates. I wouldn't worry about that.

cecton commented:
This macro looks nice btw. It emulates the deconstructor of JS:

```rs
    #[test]
    fn imap_deconstruct() {
        let my_imap = [(IString::from("foo"), 1), (IString::from("bar"), 2)]
            .into_iter()
            .collect::<IMap<IString, u32>>();
        imap_deconstruct!(
            let { foo, bar, baz } = my_imap;
            let { foobarbaz } = my_imap;
        );
        assert_eq!(foo, Some(1));
        assert_eq!(bar, Some(2));
        assert_eq!(baz, None);
        assert_eq!(foobarbaz, None);
    }
```

cecton commented:
Finally found some time to finish this :sweat: 

@hamza1311 this is ready for review! I added some website doc in "advanced topics". I published imut but I renamed it to implicit-clone. I found the name more explicit (no pun intended). It's also fully documented and I can add some people as owner on the repo so if Yew's team need to update something and I'm not available, you will have access.

hamza1311 commented:
Considering how much it goes against Rust's way of doing things, I don't think we should make this a part of Yew. One idea I have is to move the immutable example to your crate and add some docs for it in the community tab.

I want Yew to stay unopinionated about this. If someone wants to pass MBs of a Vec down as props, all the power to them. It's good to give the option to use immutable data structures. I'd prefer if this option was given as purely an external library. I had removed examples which were there only to show community libraries when I refactored services and format module. I would like to keep it that way.

Also as for where should `ImplictClone` go, we can have the following if really needed:

```rust
// yew
pub trait ImplicitClone: Clone { ... }
```
```rust
// implicit-clone
pub trait ImplicitClone: Clone { ... }

#[cfg(feature = "yew")]
impl<T> ImplicitClone for T
where T: yew::html::ImplicitClone 
{ ... }
```

cecton commented:
Using immutables is a common pattern in react. It's less used nowadays because of react hooks but it doesn't make it invalid. The very nature of having to pass properties down to children is core to this kind of framework so cheap clone is necessary. What do you think happens in react when you pass arrays to a child component: the array is not cloned, nor its values, the array is just passed as is and garbage collected like everything. That is why the golden rule is to never mutate props because it can mess with parent components.

> Considering how much it goes against Rust's way of doing things, I don't think we should make this a part of Yew.

I opened a discussion prior to this PR to avoid getting into this situation.

I made a draft PR before documenting and polishing.

There is no way turning back for me now.

> Also, if we re-export `immutable` from Yew, why shouldn't we re-export say, `gloo`?

Honestly I don't know. I assume the author wanted to give the user the choice. Personally I think we pay this power of choice with extra complexity and maintenance. Using only gloo would be a valid option imo but I don't know much about this subject.

All I know is being opionated for a framework is not necessarily a bad thing because it simplifies things.

hamza1311 commented:
I agree with being being opinionated for a framework. 

> I opened a discussion prior to this PR to avoid getting into this situation.
> 
> I made a draft PR before documenting and polishing.
> 
> There is no way turning back for me now.

I think you misunderstood what I said. I like what you've made and how you've polished it up. It should exist as-is. The only change I suggested is integrating it deeply in Yew. I guess it's also fine if we integrate this kind of immutability in Yew. I just don't want it to _the only_ option available.  
I suggested that:
- `implicit-clone`'s implementation will remain as it is right now
- `ImplictClone` trait will be added back in Yew
- `implicit-clone` will add a new feature that adds a blanket implementation of `implicit_clone::ImplicitClone` for `yew::ImplicitClone`
- `immutable` example will move to `implicit-clone`'s repo

If you want to keep it as it is, that's also fine by me. If we decide to keep it, this API looks good to me. An example demonstrating it's use with hooks will be great to have though.

CC: @futursolo @intendednull, can you comment on how/what effects it has with the state management aspects - especially in context of `bounce` and `yewdux`?

intendednull commented:
This is a cool idea, and I can see the usefulness. However I do agree it should not be the default for Yew. 

Personally don't mind the extra complexity of cloning manually. I'm really lazy, so it helps me consider all options, and try to find simpler ways of doing things first.

Opinionated features for large frameworks should typically be opt-in, and are usually kept to external crates. Yewdux, for example, started as a PR here, but I quickly realized there were many ways of solving state management, each with their own cost/benefits. Integrating it into the core library would have been bad for all the people that disagree with my method.

To answer @hamza1311 , it should have minimal to no affect on Yewdux. Might even support this as a feature if there's interest.

futursolo commented:
There should be no impact for `bounce` as well. `bounce` does not use the `ImplicitClone` trait or pass states via properties.

Personally, I think this pull request is targeting the issue with specialisation instead of generalisation.

Specialised data types are needed for JavaScript as `Object.freeze` is not deep freeze.
So if native data types are used, everytime a new item is added to the collection type,
the code needs to deep freeze the new item recursively.

In Rust, variables are immutable by default and immutability is transitive by default.
So you can achieve the same guarantee by simply wrapping the type in an `Rc` or passing an immutable reference.

Rust has a built-in type `Cow`, which is an enum with 2 variants: `Borrowed` and `Owned`.
If clone is called on an owned variant, the underlying value will be cloned.

If we have a type that replaces the Owned variant with an `Rc`'ed variant,
when clone is called on the reference counted variant, Rc will be cloned instead of the value itself.
`ImplicitClone` can then be implemented on the new Ref or Rc type.
`IString` and `IArray` can then become simple type alias.

This should allow any type wrapped by the ref or rc type to gain implicit clone in the same fashion.

For dictionaries, the situation is a little bit different.
If `IMap`'s static variant is also a real Hash Map, lookups would be `O(1)`.
So I think if `IMap` is also trying to provide random access, `Rc<IndexMap<K, V>>` is actually better than `IMap`
when combined with `std::lazy::Lazy`(`once_cell::unsync::Lazy`).

cecton commented:
> I want Yew to stay unopinionated about this. If someone wants to pass MBs of a Vec down as props, all the power to them.

I think the main problem of this PR and this feature is that I failed to explain properly why this is so important.

Please bare with me. I would like to give this another shot.

Immutable are not meant to be useful for people who want to send megabytes in a Vec. Immutable's main purpose in Yew is two folded:

1. Properties in Yew should be easy to pass down to children. The perfect example of this is AttrValue and ImplicitClone which allows the user to write something like `<MyComponent stuff="thing" />` but also like this `<MyComponent stuff=format!("thing {}", index) />`. (This was possible before the introduction of ImplicitClone and AttrValue but only for Yew core components. It was impossible to do that for https://yewprint.rm.rs components)
2. Properties in Yew must be cheap to clone. Everytime you want to add a new property to your component you need to make sure the user of your component can cheaply clone it or create a new instance easily. You can see this with AttrValue too: if you have multiple levels of component and you pass the exact same string down the line, every component's `view()` this string goes through will increase the count on the Rc but it won't cost re-instantiation.

How does react handles this? Answer: it doesn't!! All the values in JS are garbage collected so when you pass a string or an array or a map you don't clone them, you just pass another reference. This is very cheap and it's transparent for the user: "they just pass the array/string/map and it works".

With this out of the way, I hope you understand better how react-like framework is tightly coupled to a garbage collector. Yew being a react-like framework for Rust has a pretty unique problem: no garbage collection.



> Considering how much it goes against Rust's way of doing things, I don't think we should make this a part of Yew.

AttrValue is literally an immutable string cheap to clone. This is currently in Yew because it is core feature because we need to deal with no garbage collector. I assume you don't want this to get out of Yew right?

So let's see the other immutable types this PR is introducing:

1. immutable array:

This one should really be no surprise as it is the generalization of immutable string. AttrValue is `Rc<str>` rc'ed string slice. while IArray is `Rc<[T]>` rc'ed slice of T.

The question is: why would anyone want to pass an array by props to children components? Check blueprint and yewprint components. Simple things like select components use this.

What does it cost to make a generic implementation for you? It's equivalent to AttrValue. Only std stuff. There aren't 100x ways to create IArray.

2. immutable map:

This one is more complicated because there is no primitive for "map slice" in rust. The implementation in imut (now called implicit-clone) is behind a feature flag and is marked as not stable and subject to change because of that.

What is the cost of IMap? It has one dependency on IndexMap (which Yew already depends on!). Other than that it's similar to the others.

Do we need this in Yew? Honestly I don't have the answer yet here. I don't mind to put this behind a feature flag but again it's a cheap simple type. The answer is complex because react relies on these maps because it is their objects. In rust these are structs. But sometimes you do want to have a dynamic map. For a select component for example you need to provide options which are basically pairs like `(key, label)`.

> * `implicit-clone` will add a new feature that adds a blanket implementation of `implicit_clone::ImplicitClone` for `yew::ImplicitClone`

I tried that already but it didn't work because of specialization I think.

Note also that my original idea was to make yew-immutable but then in the discussion you made the case for generalizing immutable types so it's not specific to Yew and I got convinced. In order to do that you need to turn this immutable thingy (ImplicitClone and IString, IArray, IMap) to a crate that yew's core will depend on. I know it's scary, that is why I propose to add as many Yew maintainers as possible.

The big "but why?" for you I think is why ImplicitClone has to move? Why not keep it in Yew?

It's two folded again:

1. in Yew, ImplicitClone means the user don't need to write `.clone()` to pass it as props. This is huge in term of usability for Yew because of it's react-like nature
2. in term of immutability: an object that is cheap to clone can be cloned implicitly. This is huge also because it means that instead of dealing with references and lifetimes, you can deal with owned type directly. Exactly like objects that implement Copy

I need this for IArray and IMap because they are collections of things that implement ImplicitClone themselves. For instance, iterating over IArray does not yields references, it yields clones of the elements directly. It definitely is an anti pattern for Rust but it is tremendously helpful in a react-like framework where garbage collection is required.

Let me show you my very first yewprint component. The Tree component: https://yewprint.rm.rs#tree

The user needs to pass a tree-like structure. Then yewprint's tree component will render and do all the magic.

Obviously a tree-like structure is not std and definitely not a primitive. Whatever it is, we need to wrap it in Rc to make it cheap to clone.

For this to tree-like structure I did not reinvent the wheel, I used the crate id-tree. Then I made a struct to wrap it in an Rc so it can be cheaply cloned every time the parent component render the tree component:

```
struct MyComponent {
    tree: TreeDataStuff,
}

impl Component for MyComponent {
    // ...
    fn view(&self) -> Self {
        html! {
            // ....
            <Tree tree=self.tree.clone() /> // MUST be cheap
        }
}
```

My point is: cheap cloning & immutability of props is core to Yew and react-like frameworks in general. It's a particular challenge in Rust because garbage collection is not rooted in the language like others. String, arrays and (probably) maps are essential data structures that need to have a basic common implementation for every yew library (like yewprint).

cecton commented:
> I agree with being being opinionated for a framework.
> 
> > I opened a discussion prior to this PR to avoid getting into this situation.
> > I made a draft PR before documenting and polishing.
> > There is no way turning back for me now.
> 
> I think you misunderstood what I said. I like what you've made and how you've polished it up. It should exist as-is. The only change I suggested is integrating it deeply in Yew. I guess it's also fine if we integrate this kind of immutability in Yew. I just don't want it to _the only_ option available. I suggested that:
> 
>     * `implicit-clone`'s implementation will remain as it is right now
> 
>     * `ImplictClone` trait will be added back in Yew
> 
>     * `implicit-clone` will add a new feature that adds a blanket implementation of `implicit_clone::ImplicitClone` for `yew::ImplicitClone`
> 
>     * `immutable` example will move to `implicit-clone`'s repo
> 
> 
> If you want to keep it as it is, that's also fine by me. If we decide to keep it, this API looks good to me. An example demonstrating it's use with hooks will be great to have though.
> 
> CC: @futursolo @intendednull, can you comment on how/what effects it has with the state management aspects - especially in context of `bounce` and `yewdux`?

I'll do that

cecton commented:
> An example demonstrating it's use with hooks will be great to have though..

I'll do that

WorldSEnder commented:
I just had a closer look, and it seems that while it implements immutable data structures, I lot of the time, especially with arrays and maps, one wants to actually use _persistent_ datastructures, where updates are also cheap, not only clones.

As @hamza1311, I'd also favor keeping `ImplictClone` in yew. Assuming that, most of the things make a good library that can be useful, but I don't think it's strictly needed internally.

cecton commented:
> but I don't think it's strictly needed internally.

@WorldSEnder Check the first part of this comment where I give another shot at making the case of why this is core feature https://github.com/yewstack/yew/pull/2563#issuecomment-1100587583

If you want a short answer: look at AttrValue. AttrValue is literally immutable string. Cheap to clone. This PR extends it to arrays and map.

> a lot of the time, especially with arrays and maps, one wants to actually use _persistent_ datastructures, where updates are also cheap, not only clones.

@WorldSEnder I'm not sure I'm following. :thinking: this PR introduces immutable types to immutable data (not updatable) that can be passed in props. Props are not supposed to be mutated ever. I think you have something else than props in mind.

> In Rust, variables are immutable by default and immutability is transitive by default. So you can achieve the same guarantee by simply wrapping the type in an `Rc` or passing an immutable reference.
> 
> Rust has a built-in type `Cow`, which is an enum with 2 variants: `Borrowed` and `Owned`. If clone is called on an owned variant, the underlying value will be cloned.

@futursolo look at AttrValue. It has been introduced exactly to replace the Cow that was there before. And I did not personally introduce it. AttrValue is literally an immutable string. I even aliased the type here in this PR because it's 100% identical.

This is a larger discussion here about the reason why we have introduced AttrValue in the first place but for yewprint it is a huge deal. This is a needed feature.

> Opinionated features for large libraries should typically be opt-in, and are usually kept to external crates. Yewdux, for example, started as a PR here, but I quickly realized there were many ways of solving state management, each with their own cost/benefits. Integrating it into the core library would have been bad for all the people that disagree with my method.

@intendednull This is :100: correct.

It's a golden rule **but not a silver bullet**. You can't have everything as opt-in feature. There are valid cases where you need to make an opinionated choice for the library. I believe mutability is one of them. Looking at AttrValue, which is literally an immutable string, you can see there is a need for immutability. This PR only introduces 2 more types: immutable arrays and maps.

Is it cheap to add to Yew's core: the answer is definitely yes. You can see it in the wasm size check of the examples. IArray only uses std and has very little code. IMap is slightly questionable but it is cheap to add to Yew as we already have a hard dependency on IndexMap anyway.

If you want another example of opinionated choice Yew make for its users: `Classes`. This type and macro is the Rust equivalent of the famous classnames library in node.

My point here is that you can't apply the rule of everything should be opt-in to every single feature. Some choices need to be made for legit reasons.

futursolo commented:
> @futursolo look at AttrValue. It has been introduced exactly to replace the Cow that was there before. AttrValue is literally an immutable string. I even aliased the type here in this PR because it's 100% identical.

I think I might have not explained this properly.
The following struct is a type that I implemented for [another project](https://github.com/futursolo/stylist-rs/blob/unified-parser/packages/stylist-core/src/arc_ref.rs#L15) to facilitate this purpose:

```rust
/// A smart reference that lives at least `'a`.
///
/// This type behaves like [`Cow`](std::borrow::Cow) but uses [`Arc`] to store the owned value so
/// it's cheap to clone.
#[derive(Clone)]
pub enum ArcRef<'a, T>
where
    T: ?Sized + ToOwned,
{
    Arc(Arc<<T as ToOwned>::Owned>),
    Ref(&'a T),
}
```

Then str and slice can become simple type alias.

```rust
// Can optionally made to be 'static.
type IStr<'a> = ArcRef<'a, str>;
type ISlice<'a, T> = ArcRef<'a, [T]>;
```

cecton commented:
@hamza1311 https://github.com/yewstack/yew/blob/edceb155f612de341e343a7078bec907783cffe9/examples/immutable/src/array.rs

WorldSEnder commented:
> Props are not supposed to be mutated ever. I think you have something else than props in mind.

I'm specifically referring to the example and [this use case](https://github.com/yewstack/yew/pull/2563/files#diff-6a1dcb81a13cd0312792ccc2f198fc6f3e7412d156ac151d45833aa820958fb7R45-R54). A persistent array would shorten this to something along the lines of

```rust
// IMap::insert doesn't modify, but returns a new IMap
self.values = self.values.insert(self.values.len() as u32, IString::from(name)),
```

In case its not clear what the term _persistent_ means in this context

> a persistent data structure or not ephemeral data structure is a data structure that always preserves the previous version of itself when it is modified. Such data structures are effectively immutable, as their operations do not (visibly) update the structure in-place, but instead always yield a new updated structure

The example currently constructs the new prop by collecting from an iterator of the old values, and then chaining the new one. This operation is O(n). A persistent array data structure can offer O(1) append, while still being immutable, i.e. cheap to clone. This makes it in my view even better for props than using "simple" immutable datastructures such as a wrapped `Rc<[T]>`.

> AttrValue is literally immutable string. Cheap to clone. This PR extends it to arrays and map.

Yep, I agree that the concept of `AttrValue` (or even something more powerful) can be found in [mutliple](https://github.com/bnclabs/ppar) + [other](https://github.com/orium/rpds) + [immutable](https://docs.rs/im/latest/im/#lists) libraries - now also in `imut`. If I had confidence that one of them could replace all other uses for most users, I'd favor switching to it. What I'm arguing is that it's not clear which library will emerge here and deeply integrating `imut` at this point is a bit rushed.

WorldSEnder commented:
> If you want another example of opinionated choice Yew make for its users: `Classes`. This type and macro is the Rust equivalent of the famous classnames library in node.

You are on to something with `Classes`. It's currently doesn't have an immutable design and caching it or using it as a prop leads to pointless clones all over the place. Reconstructing it on every render seems the way to go and at this point it's no cheaper than cloning it everywhere. I think this could benefit from a persistent DS approach as an `IList<String>`.

cecton commented:
> Then str and slice can become simple type alias.

I just tested (for IString) and it seems to work nicely. Excellent! :exploding_head: 

There is a difference though as it is storing a `Rc<String>` instead of `Rc<str>` for the owned type... https://doc.rust-lang.org/std/primitive.str.html#impl-ToOwned

I wonder if this could be avoided using the From impls on Rc... :thinking: I will investigate more.

cecton commented:
> You are on to something with `Classes`. It currently doesn't have an immutable design and caching it or using it as a prop leads to pointless clones all over the place. Reconstructing it on every render seems the way to go and at this point it's no cheaper than cloning it everywhere. I think this could benefit from a persistent DS approach as an `IList<String>`.

heh ... You're right. It is actually partly cheap to clone as it is using `IndexSet<Cow<'static, str>>` so static str will be cheap but owned string will be cloned fully. It's probably worth to convert this to `IArray<IString>` in this case. Thanks it's really good catch.

> I'm specifically referring to the example and [this use case](https://github.com/yewstack/yew/pull/2563/files#diff-6a1dcb81a13cd0312792ccc2f198fc6f3e7412d156ac151d45833aa820958fb7R45-R54). A persistent array would shorten this to something along the lines of
> 
> ```rust
> // IMap::insert doesn't modify, but returns a new IMap
> self.values = self.values.insert(self.values.len() as u32, IString::from(name)),
> ```

This API is actually very alike to immutable.js' API :thinking: ... I did remove this from my checklist because I'm afraid it would be conflicting or at least confusing. But in this example you just gave it's very handy compared to collecting from an iterator.

It might be worth giving a shot.

> In case its not clear what the term _persistent_ means in this context
> 
> > a persistent data structure or not ephemeral data structure is a data structure that always preserves the previous version of itself when it is modified. Such data structures are effectively immutable, as their operations do not (visibly) update the structure in-place, but instead always yield a new updated structure
> 
> The example currently constructs the new prop by collecting from an iterator of the old values, and then chaining the new one. This operation is O(n). A persistent array data structure can offer O(1) append, while still being immutable, i.e. cheap to clone. This makes it in my view even better for props than using "simple" immutable datastructures such as a wrapped `Rc<[T]>`.

Oh I see know. Ok so I assume we would need to store which version holds which element.... :thinking: that will probably make slicing impossible. In the current implementation of IArray you can do `.as_slice()` to get a slice to the internal array. It's also useful because you it gives you Deref on slice: https://doc.rust-lang.org/std/vec/struct.Vec.html#deref-methods-%5BT%5D

I'm not sure if this can be implemented easily but it's definitely a nice idea.

> Yep, I agree that the concept of `AttrValue` as an immutable String (or even something more powerful) can be found in [mutliple](https://github.com/bnclabs/ppar) + [other](https://github.com/orium/rpds) + [immutable](https://docs.rs/im/latest/im/#lists) libraries - now also in `imut`. If I had confidence that one of them could replace all other uses for most users, I'd favor switching to it. What I'm arguing is that it's not clear which library will emerge here and deeply integrating `imut` at this point is a maybe rushed.

To be honest I did a lot of research and couldn't find **any** immutable library. :sweat_smile: 

I took a quick look to the 3 you mention here and they're all seem more sophisticated than what I provided with imut/implicit-clone. Ah I see now, it's related to the previous point you made.

In any case. I think this might not be a problem at all:

I actually renamed `imut` to `implicit-clone` because ImplicitClone is kinda the core trait. Having it outside yew allows other libraries to implement it without having to pull Yew as dependency. The immutable types provided by implicit-clone are kinda stupid-simple versions of immutable types which we can move behind a feature flag if we want to make them opt-in. We can also make PRs on those other immutable libraries to add a feature flag to activate ImplicitClone. This will allow people to use whatever crate they want to deal with immutables.

Since the stupid-simple implementation of the immutable types are very cost effective (in term of dependency and on the wasm size), they could stay enabled by default. For example, we can add a feature flag `immutable` in yew that would activate those types. Nevertheless I think AttrValue (IString) is still required because it's used internally in Yew and, as you pointed out, we should probably do the same with IArray for Classes (though, it's an IndexSet, not an array).

cecton commented:
@futursolo this seems to work fantastically well. Though I did get away from the ToOwned trait as it involves unnecessary owned types https://github.com/rustminded/implicit-clone/pull/5

It won't work for IMap but that's okay because it's a really special case anyway.

cecton commented:
@hamza1311 @WorldSEnder @futursolo @intendednull hello :wave: is there anything I need to do to get approval for this PR?

WorldSEnder commented:
Sorry this took so long to get back to, thanks for the ping. At this point, the only concerns left are with the following 3 lines:

https://github.com/yewstack/yew/pull/2563/files#diff-3bb6b467c4323ac61d59273089829fda9f98e61340e754ecbeb76b73c9415849R46

https://github.com/yewstack/yew/pull/2563/files#diff-fbd09dd5a253cf9312c385e7abe03ea8ec90cd6926ba8bbf4512bae8372b25d1R4

less so
https://github.com/yewstack/yew/pull/2563/files#diff-6940ed8004d1a15283f03d61dd65e1bc613dc6b2b4779eb3e43860cc2a732cb6R291

These are all re-exports of `imut` in yew. I don't consider myself an expert in API-breaking changes so can someone answer the question: What if we notice we want to do things slightly different than imut, would it be a breaking change to revert those back to internal types/traits?

I don't have a problem with giving `imut` first class support for some `IntoPropValue` and similar, but for stability, would it make sense to write `AttrValue` as a wrapper of `IString` instead of an alias and keep the `ImplicitClone` (like the `IntoPropValue`) as a trait in yew (giving impls for the types in imut somewhere)?

cecton commented:
> These are all re-exports of `imut` in yew. I don't consider myself an expert in API-breaking changes so can someone answer the question: What if we notice we want to do things slightly different than imut, would it be a breaking change to revert those back to internal types/traits?

Yes it's a breaking change to do so because the trait and structs will belong to a different crate. But imut (now called implicit-clone) is made by myself mainly for Yew. I can add any owner to this project. So I don't expect things to diverge. Though, the hope is still to keep this as generic as possible.


> I don't have a problem with giving `imut` first class support for some `IntoPropValue` and similar, but for stability, would it make sense to write `AttrValue` as a wrapper of `IString` instead of an alias and keep the `ImplicitClone` (like the `IntoPropValue`) as a trait in yew (giving impls for the types in imut somewhere)?

It was actually @hamza1311's idea to bring this more generic so it can be used elsewhere and not just for Yew. I can make a crate yew-immutable instead that would enrich Yew's `ImplicitClone`, that was what I did in the beginning. See discussion #2448. But in the end I agreed that it is actually better to have this more generic.

hamza1311 commented on packages/yew/src/lib.rs:
```diff
@@ -288,6 +288,7 @@ mod server_renderer;
 pub mod suspense;
 pub mod utils;
 pub mod virtual_dom;
+pub use implicit_clone::unsync as immutable;
```
We shouldn't re-export it. We don't do it for `gloo` and we removed the re-export for web-sys: #2084. `implicit_clone` should also not be re-exported. Any types consumed in the public API can/should be re-exported as needed.

cecton replied:
Ok makes sense

hamza1311 commented on website/docs/advanced-topics/immutable.mdx:
```diff
@@ -0,0 +1,84 @@
+---
+title: 'Immutable'
+description: 'Immutable.js-like types for Yew'
```
```suggestion
title: 'Immutable Types'
description: 'Immutable data structures for Yew'
```

hamza1311 commented on website/docs/advanced-topics/immutable.mdx:
````diff
@@ -0,0 +1,84 @@
+        <ul>
+        { for props.folks.iter().map(|s| html!(<li>{s}</li>)) }
+        </ul>
+        </>
+    }
+}
+
+#[function_component(ArrayExample)]
+pub fn array_example() -> Html {
+    let folks = use_state(IArray::<IString>::default);
+    let onkeyup = {
+        let folks = folks.clone();
+        Callback::from(move |e: KeyboardEvent| {
+            if e.key() == "Enter" {
+                let event: Event = e.dyn_into().unwrap_throw();
+                let event_target = event.target().unwrap_throw();
+                let target: HtmlInputElement = event_target.dyn_into().unwrap_throw();
+                let name = target.value();
+                target.set_value("");
+
+                folks.set(
+                    folks
+                        .iter()
+                        .chain(std::iter::once(IString::from(name)))
+                        .collect(),
+                );
+            }
+        })
+    };
+
+    html! {
+        <>
+        <h2>{"Input"}</h2>
+        <input {onkeyup} />
+        <h2>{"Output"}</h2>
+        <FolksView folks={&*folks} />
+        </>
+    }
+}
+```
````
There's no need to show all that code in a documentation page. A complete example should be in the examples directory

cecton replied:
Good point

hamza1311 reviewed (changes requested):
With regards to breaking changes, it's the same as if a new version to Gloo were to be published, with changes to it's public API. That would be a breaking change for Yew. We don't have to worry about breaking changes unless we update the dependency. This is why only the types used in public API should be re-exported.

> But imut (now called implicit-clone) is made by myself mainly for Yew. I can add any owner to this project. So I don't expect things to diverge. Though, the hope is still to keep this as generic as possible.

It would be great if you could transfer the repository to this organization. I don't see it's usage being much different from gloo.

Since we're moving `ImplicitClone` to it's own library, I'm fine with adding first-class for support for this library. We used other libraries like this too, such as `tokio`, `index-map`, etc

cecton commented:
> It would be great if you could transfer the repository to this organization. I don't see it's usage being much different from gloo.

oh wow I wanted to send the invitation to take the repository but because I'm admin on yewstack it got transferred without approval. So... ok it's done :sweat_smile:

cecton commented:
The unit test in 1.56 is failing the same on main so I suppose this error does not come from this PR.

@hamza1311 I applied the suggestions you asked

WorldSEnder commented:
The unit tests should be fixed now, on master. Can you merge and push to re-trigger them?

cecton commented:
@WorldSEnder done :+1: all tests are green

hamza1311 commented on website/docs/advanced-topics/immutable.mdx:
```diff
@@ -0,0 +1,24 @@
+
+## Why using immutable types?
+
+In a React-like framework, properties are propagated from ancestors to
```
```suggestion
Properties are propagated from ancestors to
```

cecton replied:
Fine by me but I'm afraid that if we try to distance ourselves too much from "being a clone of react" we will lose the understanding.

Properties are propagated from ancestors to children BECAUSE it's how these kind of frameworks work. On the other hand, the "word" property can have a completely different meaning in other frameworks. For instance, in VisualBASIC (I haven't work on GUI for a long time lol), "properties" meant the thingies you set on the objects and activate/deactivate a specific feature. There is no notion of propagation at all.

I'm also personally convinced that people who start with Yew are looking for React in Rust. The documentation should help them navigate through it.

hamza1311 replied:
> I'm also personally convinced that people who start with Yew are looking for React in Rust.

It was quite the opposite for me. I learnt Yew before I learnt React. I believe this should be better a way to word this:
```suggestion
Properties, like in React, are propagated from ancestors to
```

I'm kind of hesitant of documentation making Yew sound like it's a React clone or (more importantly) making React be the be-all and end-all of web frameworks/libraries. It is inspired by React but there are certain areas where we shouldn't (or just can't) following what React does because Rust is a entirely a different language and requires a different mindset to work with. It's not the case here though.

cecton replied:
> It is inspired by React but there are certain areas where we shouldn't (or just can't) following what React does because Rust is a entirely a different language and requires a different mindset to work with.

:100: I totally agree with you. Definitely not the same languages and problems are often solved very differently. I'm also discussing about this here too: https://github.com/yewstack/yew/issues/1843#issuecomment-1157240010

> I'm kind of hesitant of documentation making Yew sound like it's a React clone or (more importantly) making React be the be-all and end-all of web frameworks/libraries

I think you misunderstood me. I never said React is the be-all and end-all of anything.

My point is that React is one huge piece of influence in web development for years and people arrive in Rust and in Yew with already years of experiences in React. It seems it is not your case (good for you actually) but for the rest of us we need to unlearn some things and adapt to the Rust reality.

cecton replied:
(I kept the conversation "unresolved" so it is not hidden in the PR but I already committed your proposal with the ", like in React,")

hamza1311 commented on packages/yew/Cargo.toml:
```diff
@@ -29,6 +29,7 @@ thiserror = "1.0"
 
 futures = { version = "0.3", optional = true }
 html-escape = { version = "0.2.9", optional = true }
+implicit-clone = { version = "0.2", features = ["map"] }
```
Do we need to enable `map` feature? I don't think we use it.

cecton replied:
No... I added it because there was the re-export and it felt like a core feature. But on the other hand, it's an invitation to use an anti-pattern that is not even useful for Yew. It's maybe best to remove it entirely from implicit-clone too

hamza1311 replied:
I agree that we should remove it:

```suggestion
implicit-clone = { version = "0.2" }
```

cecton replied:
What do you think about removing it from implicit-clone too? You're also maintaining a UI framework so I guess you probably met a lot of use cases where we need to pass lists in properties and maybe different types.

On my side I didn't find any use of passing maps. The only component that does use a more complex structure for state is the [tree](https://yewprint.rm.rs/#tree) but I opted for using a specialized tree library instead. The ones that use Vec will greatly benefit from IArray.

WorldSEnder replied:
`Attributes` is more-or-less a map, although it's more specialized I guess. Overall, it makes sense to have such a thing in implicit-clone, though I confess I haven't looked at the implementation what you're currently using to implement it.

cecton replied:
Ah yes I forgot there is that Attributes thingy...

Well... yes probably Attributes is the same (at least same idea) than IMap but it shouldn't be a concern here. Attributes is used for internal machinery in Yew. Here we are talking about API and what tools we want to expose to the user. I don't know if anyone needs Attributes? Do they? Oh... it's public API... haha okay so maybe :grin: why is this public API?? xD (cc @siku2 maybe you know something? I think you've been maintaining Yew the longest)

cecton replied:
@hamza1311 I forgot to mention this but the map feature brings zero dependencies as it depends on indexmap which yew depends on anyway.

...come to think of it... :thinking: even if I didn't find a use in my use cases, it seems that having an immutable map collection is still something useful. IMap has 2 advantages:
1. it implements ImplicitClone: this means all its API yields clones instead of references which is very useful for properties
2. you can create static collections (for small collections)

I still removed the feature in the dependencies because it's optional anyway. I did it because I originally re-exported but now that we don't re-export there is absolutely no point. (tbh... I really don't get why we ban re-export :woman_shrugging:)

@WorldSEnder I just checked Attributes and it has quite some big differences:
- Attributes can only hold strings vs IMap can hold anything that implements ImplicitClone itself: it could be faster to use Attributes
- Attributes has a variant `Dynamic` which allow using subslices of an existing set of keys/values": it's probably used to optimize Yew somewhere... I'm not sure if IMap should also implement this, I don't see much value tbh.

Overall I have the feeling that Attributes is really specialized for Yew. The reason why it is public API is because it is used by yew-macro so it has to be.

cecton replied:
@hamza1311 oh nevermind what I said. We have to keep the feature map otherwise I can't make the impls for IntoPropValue which is a yew trait. Chicken and egg problem.

Unless there is a way to put a `#[cfg()]` on the features of a child dependency, I don't see any other way. It's not a big deal anyway since it brings no dependencies.

hamza1311 replied:
It doesn't matter either way. It's fine that you removed it. Keeping it around wouldn't have mattered either

hamza1311 reviewed (commented):
This can be merged after #2740 because otherwise CI breaks

cecton commented:
I'm not sure what I need to do to fix the tests :woman_shrugging:

hamza1311 commented:
> I'm not sure what I need to do to fix the tests :woman_shrugging:

I fixed them

hamza1311 reviewed (approved):
Finally! We're ready to merge this in

cecton commented:
Thanks a lot for the fixes @hamza1311 :hugs: 

And thanks everybody for all your time on the matter! :heart:

martinitus commented:
This seemed to be a pretty substantial change to yew, would be awesome to read about it somewhere in the "next" docs  :)
Was really nice reading through this :)

cecton commented:
Oh sorry I think the doc on that will be very succinct. It is worth a blog post in my opinion but every time I tried to explain the theory I feel like people did not get it. Maybe a post with an example that shows in which order the components are updated and when the memory is freed would be the best way to understand the concept.

To summarize, it's two-folded:

1. React-like frameworks (components updating from root (top) to child (bottom)) work great when you have a garbage collector... but it's terrible in Rust because the child properties must outlive its parent properties. Very very succinct right? xD

2. The other issue immutables is solving is for Yew libraries that require to pass arrays, strings and maps in properties because it can reduce the amount of allocations. Example of component that uses a list: the drop list https://yewprint.rm.rs/html-select Allocating vecs in the `fn view()` is not great at all.

hamza1311 commented:
> It is worth a blog post in my opinion

@cecton do you want to write on up?

cecton commented:
@hamza1311 sure! xD There are so many things I say I will/would do. I should stop saying things I would do and start doing haha I will put that on my TODO list
